diff --git a/arch/arm/dts/imx8mq-var-dart.dts b/arch/arm/dts/imx8mq-var-dart.dts
index 2646ccb..0d70fb7 100644
--- a/arch/arm/dts/imx8mq-var-dart.dts
+++ b/arch/arm/dts/imx8mq-var-dart.dts
@@ -54,7 +54,7 @@
 	phy-reset-gpios = <&gpio1 9 GPIO_ACTIVE_LOW>;
 	phy-reset-duration = <10>;
 	fsl,magic-packet;
-	status = "okay";
+	status = "disabled";
 
 	mdio {
 		#address-cells = <1>;
@@ -122,10 +122,7 @@
 };
 
 &usb_dwc3_0 {
-	dr_mode = "otg";
-	hnp-disable;
-	srp-disable;
-	adp-disable;
+	dr_mode = "host";
 	usb-role-switch;
 	snps,dis-u1-entry-quirk;
 	snps,dis-u2-entry-quirk;
@@ -133,12 +130,12 @@
 };
 
 &usb3_phy1 {
-	status = "okay";
+	status = "disabled";
 };
 
 &usb_dwc3_1 {
 	dr_mode = "host";
-	status = "okay";
+	status = "disabled";
 };
 
 /* eMMC */
diff --git a/board/variscite/imx8mq_var_dart/imx8mq_var_dart.c b/board/variscite/imx8mq_var_dart/imx8mq_var_dart.c
index 118fb56..3df5050 100644
--- a/board/variscite/imx8mq_var_dart/imx8mq_var_dart.c
+++ b/board/variscite/imx8mq_var_dart/imx8mq_var_dart.c
@@ -145,6 +145,27 @@ int board_usb_cleanup(int index, enum usb_init_type init)
 }
 #endif
 
+static iomux_v3_cfg_t const agr_pads[] = {
+        IMX8MQ_PAD_NAND_DATA03__GPIO3_IO9 | MUX_PAD_CTRL(NO_PAD_CTRL),
+        IMX8MQ_PAD_NAND_DATA05__GPIO3_IO11 | MUX_PAD_CTRL(NO_PAD_CTRL),
+};
+
+#define LCD_EN IMX_GPIO_NR(3, 9)
+#define BLOWER_EN IMX_GPIO_NR(3, 11)
+
+int agr_cfg_init(void)
+{
+        imx_iomux_v3_setup_multiple_pads(agr_pads, ARRAY_SIZE(agr_pads));
+        gpio_request(LCD_EN, "LCD_EN");
+        gpio_direction_output(LCD_EN, 1);
+        gpio_request(BLOWER_EN, "BLOWER_EN");
+        gpio_direction_output(BLOWER_EN, 1);
+
+        return 0;
+}
+
+
+
 int board_init(void)
 {
 #ifdef CONFIG_FEC_MXC
@@ -155,6 +176,8 @@ int board_init(void)
 	init_usb_clk();
 #endif
 
+	agr_cfg_init();
+
 	return 0;
 }
 
diff --git a/include/configs/imx8mq_var_dart.h b/include/configs/imx8mq_var_dart.h
index 1b95120..030a011 100644
--- a/include/configs/imx8mq_var_dart.h
+++ b/include/configs/imx8mq_var_dart.h
@@ -45,6 +45,9 @@
 
 #define CONFIG_REMAKE_ELF
 
+#undef CONFIG_CMD_NET
+#undef CONFIG_FEC_MXC
+
 /* ENET Config */
 /* ENET1 */
 #if defined(CONFIG_FEC_MXC)
@@ -80,8 +83,9 @@
 	"mmcautodetect=yes\0" \
 	"mmcpart=1\0" \
 	"m4_addr=0x7e0000\0" \
-	"m4_bin=hello_world.bin\0" \
-	"use_m4=no\0" \
+	"m4_bin=agronave_m4.bin\0" \
+	"m4_len=0x80000\0" \
+	"use_m4=yes\0" \
 	"loadm4bin=load mmc ${mmcdev}:${mmcpart} ${loadaddr} ${bootdir}/${m4_bin} && " \
 	"cp.b ${loadaddr} ${m4_addr} ${filesize}\0" \
 	"runm4bin=" \
@@ -116,11 +120,7 @@
 		"fi;\0" \
 	"findfdt=" \
 		"if test $fdt_file = undefined; then " \
-			"if gpio input 12; then " \
-				"setenv fdt_file imx8mq-var-dart-cb12.dtb; " \
-			"else " \
-				"setenv fdt_file imx8mq-var-dart.dtb;" \
-			"fi; " \
+			"setenv fdt_file imx8mq-var-dart.dtb;" \
 		"fi; \0" \
 	"loadfdt=run findfdt; " \
 		"echo fdt_file=${fdt_file}; " \
